integration/spark/src/main/common/java/io/openlineage/spark/agent/lifecycle/SparkSQLExecutionContext.java

```diff
@@ -61,29 +61,64 @@
   private final OpenLineage openLineage =
       new OpenLineage(OpenLineageClient.OPEN_LINEAGE_CLIENT_URI);
 
+  private AtomicBoolean started = new AtomicBoolean(false);
+  private AtomicBoolean finished = new AtomicBoolean(false);
+  private Optional<Integer> jobId = Optional.empty();
+
   public SparkSQLExecutionContext(
       long executionId,
       OpenLineageContext sparkContext,
+      QueryExecution queryExecution,
       List<QueryPlanVisitor<LogicalPlan, OpenLineage.OutputDataset>> outputDatasetSupplier,
       List<QueryPlanVisitor<LogicalPlan, OpenLineage.InputDataset>> inputDatasetSupplier) {
     this.executionId = executionId;
     this.sparkContext = sparkContext;
-    this.queryExecution = SQLExecution.getQueryExecution(executionId);
+    this.queryExecution = queryExecution;
     this.outputDatasetSupplier = outputDatasetSupplier;
     this.inputDatasetSupplier = inputDatasetSupplier;
     this.jobMetrics = JobMetricsHolder.getInstance();
   }
 
-  public void start(SparkListenerSQLExecutionStart startEvent) {}
+  public void start(SparkListenerSQLExecutionStart startEvent) {
+    log.debug("SparkListenerSQLExecutionStart - executionId: {}", startEvent.executionId());
+    startEvent(startEvent.time());
+  }
 
-  public void end(SparkListenerSQLExecutionEnd endEvent) {}
+  public void end(SparkListenerSQLExecutionEnd endEvent) {
+    log.debug("SparkListenerSQLExecutionEnd - executionId: {}", endEvent.executionId());
+    // TODO: can we get failed event here?
+    // If not, then we probably need to use this only for LogicalPlans that emit no Job events.
+    // Maybe use QueryExecutionListener?
```

Comment:
In Spark 3.x, the `SparkListenerSQLExecutionEnd` event has an `executionFailure` property - https://github.com/apache/spark/blob/v3.1.2/sql/core/src/main/scala/org/apache/spark/sql/execution/ui/SQLListener.scala#L68-L69 . The `QueryExecutionListener` is now just a different interface that responds to the same `SparkListenerSQLExecutionEnd` events - https://github.com/apache/spark/blob/v3.1.2/sql/core/src/main/scala/org/apache/spark/sql/util/QueryExecutionListener.scala#L139-L158 .